### PR TITLE
implement multiple working points in ID for e/g triggers based on HGCAL clusters

### PR DESCRIPTION
**Pull request by jbsauvan**
#### PR description:
Add the possibility to define several working points in the identification BDTs for the L1 e/g triggers based on HGCAL clusters.
FYI @cerminar 

#### PR validation:
Private tests in `L1Trigger/L1THGCalUtilities/test`:
```
cmsRun testHGCalL1T_RelValV11_cfg.py
cmsRun testHGCalL1T_multialgo_V11_cfg.py
```
Tested D49, D60, D68, D77, D86 workflows
```
runTheMatrix.py -w upgrade -l 23234.0 --maxSteps=2 
runTheMatrix.py -w upgrade -l 28234.0 --maxSteps=2 
runTheMatrix.py -w upgrade -l 31434.0 --maxSteps=2
runTheMatrix.py -w upgrade -l 35034.0 --maxSteps=2
runTheMatrix.py -w upgrade -l 38634.0 --maxSteps=2
```

### CODE REVIEW

**Comment by AdrianoDee** [comment]
test parameters:

- workflows = 23234.0,28234.0,31434.0,35034.0,38634.0
- relvals_opt= -w standard,highstats,pileup,generator,extendedgen,production,upgrade,cleanedupgrade,ged

**Comment by AdrianoDee** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by AdrianoDee** [comment]
+upgrade
- tests run with no issue or difference;
- implement multiple wps through a simple `WorkingPoint` class.

**Comment by jbsauvan** [comment]
Good point, thanks @perrotta 
I've changed it as proposed

File: L1Trigger/L1THGCal/plugins/backend/HGCalTriggerClusterIdentificationBDT.cc [hunk 1]
```diff
@@ -102,25 +102,26 @@ void HGCalTriggerClusterIdentificationBDT::initialize(const edm::ParameterSet& c
...
     std::string wp_name = wp_conf.getParameter<std::string>("Name");
     std::vector<double> wps = wp_conf.getParameter<std::vector<double>>("WorkingPoint");
     working_points_names_.emplace_back(wp_name);
-    if (wps.size() != categories_etamin.size()) {
+    if (wps.size() != categories_size) {
       throw cms::Exception("HGCalTriggerClusterIdentificationBDT|BadInitialization")
           << "Inconsistent number of categories in working point '" << wp_name << "'";
     }
-    for (unsigned cat = 0; cat < categories_etamin.size(); cat++) {
+    for (unsigned cat = 0; cat < categories_size; cat++) {
```
Comment (perrotta) [inline L1Trigger/L1THGCal/plugins/backend/HGCalTriggerClusterIdentificationBDT.cc:117]:
```suggestion
    for (size_t cat = 0; cat < categories_size; cat++) {
```
(to keep the same type for `cat` and `categories_size`)

File: L1Trigger/L1THGCal/plugins/backend/HGCalTriggerClusterIdentificationBDT.cc [hunk 2]
```diff
@@ -102,25 +102,26 @@ void HGCalTriggerClusterIdentificationBDT::initialize(const edm::ParameterSet& c
...
       working_points_[cat].emplace_back(wp_name, wps[cat]);
     }
   }
 
-  categories_.reserve(categories_etamin.size());
-  bdts_.reserve(categories_etamin.size());
-  for (unsigned cat = 0; cat < categories_etamin.size(); cat++) {
+  categories_.reserve(categories_size);
+  bdts_.reserve(categories_size);
+  for (unsigned cat = 0; cat < categories_size; cat++) {
```
Comment (perrotta) [inline L1Trigger/L1THGCal/plugins/backend/HGCalTriggerClusterIdentificationBDT.cc:124]:
```suggestion
  for (size_t cat = 0; cat < categories_size; cat++) {
```
(to keep the same type for `cat` and `categories_size`)

**Review by perrotta — COMMENTED** [review]
Thank you @jbsauvan 
Since you are there, you can also use the same types for the loop index, as follows

**Comment by jbsauvan** [comment]
Thanks, done

**Comment by perrotta** [comment]
please test

**Comment by AdrianoDee** [comment]
+upgrade

**Comment by cecilecaillol** [comment]
+l1

**Comment by perrotta** [comment]
+1

**Pull request merged**